### PR TITLE
Generate typescript types from rest API docs

### PR DESCRIPTION
**Pull request by bbovenzi**
Use openapi-typescript to generate types for all rest API data. This keeps a single source of truth vs manually writing all the data endpoint types.

Also, found a bug where `map_index` was missing from the TaskInstance schema

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/www/alias-rest-types.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  if (hasQuery) variableTypes.push(`${operationPath}['parameters']['query']`);
+  if (hasBody) variableTypes.push(`${operationPath}['requestBody']['content']['application/json']`);
+
+  if (variableTypes.length === 0) return '';
+  const typeName = `${toPascalCase(operationName)}Variables`;
+  return [typeName, `export type ${typeName} = ${variableTypes.join(' & ')};`];
+};
+
+// Generate Type Aliases
+// This is recurrsivenly called to generate types for Linege's external dependencies.
```
Comment (pierrejeambrun) [inline airflow/www/alias-rest-types.js]: Typo `recurrsivenly`,  `Linege's` ?

File: airflow/www/alias-rest-types.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    // recursively call this for any externals
+    if (ts.isInterfaceDeclaration(node) && node.name?.text === 'external') {
+      node.members.forEach((external) => {
+        generateAliases(external.type, writeText, `external['${external.name.text}']`);
+      });
+    }
+  });
+};
+
+function generate(file) {
```
Comment (pierrejeambrun) [inline airflow/www/alias-rest-types.js:166]: Should we also push the licence directly, or are we counting on the pre-commit hook to do that for us ?

**Review by pierrejeambrun — COMMENTED** [review]
Looking nice. Just a few suggestions.

For the camel case naming they recommend keeping it as snake case in their readme. (project goals section) Maybe we should do that and get completely rid of camel case.

If we still want to enforce camel case, as a last resort we could use the template litéral + conditionnal type recursion, and we could define this: (add this to the generated file)
```typescript
type SnakeToCamelCase<word extends string> =
  word extends `${infer firstWord}_${infer rest}`
    ? `${firstWord}${Capitalize<SnakeToCamelCase<rest>>}`
    : word;

type KeysToCamelCase<T> = {
  [K in keyof T as SnakeToCamelCase<string & K>]: T[K]
}
```

Then we only need to wrap our generated type like so: (in the generation step)
```typescript
export type UserCollectionItem = KeysToCamelCase<components['schemas']['UserCollectionItem']>;
export type User = KeysToCamelCase<components['schemas']['User']>;
export type UserCollection = KeysToCamelCase<components['schemas']['UserCollection']>;
export type ConnectionCollectionItem = KeysToCamelCase<components['schemas']['ConnectionCollectionItem']>;
```

image

**Comment by bbovenzi** [comment]
@pierrejeambrun Good idea to add a `KeysToCamelCase`. I think we may want to switch to snake_case in the future. But that's a large refactor with a large surface area to introduce bugs so I think converting the types to camelCase is fine for now.

Also, fixed spelling errors and auto-adding the license to the generated file.

**Comment by pierrejeambrun** [comment]
Should we add the type-generation step to the static-chekc (pre-commit + CI). What I mean by that is, if someone update the API, these changes should be reflected in the React types.

And if there is a breaking change in the API spec, the front should be notified and the linting should not work.

**Review by jedcunningham — APPROVED** [review]
Looks good, though there are some static check failures to clean up.

**Comment by bbovenzi** [comment]
https://github.com/apache/airflow/pull/25167 will handle the pre-commit hook to auto-generate the type file

**Pull request merged**